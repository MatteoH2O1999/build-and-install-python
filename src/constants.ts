// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

export enum InputNames {
  PYTHON_VERSION = 'python-version',
  PYTHON_VERSION_FILE = 'python-version-file',
  ARCHITECTURE = 'architecture',
  CACHE_BUILD = 'cache-build',
  ALLOW_BUILD = 'allow-build',
  TOKEN = 'token'
}

export enum OutputNames {
  PYTHON_VERSION = 'python-version',
  ARCHITECTURE = 'architecture'
}

export enum ManifestUrl {
  OWNER = 'actions',
  REPO = 'python-versions',
  BRANCH = 'main'
}

const defaultPyPy2 = 'pypy2.7';
const defaultPyPy3 = 'pypy3.9';

export {defaultPyPy2, defaultPyPy3};

const windowsBuildDependencies = ['Microsoft.VisualStudio.Component.VC.140'];

const vsInstallerUrl = 'https://aka.ms/vs/17/release/vs_enterprise.exe';

export {windowsBuildDependencies, vsInstallerUrl};

const ubuntuDependencies = [
  'make',
  'build-essential',
  'libssl-dev',
  'zlib1g-dev',
  'libbz2-dev',
  'libsqlite3-dev',
  'libncursesw5-dev',
  'libreadline-dev',
  'libgdbm-dev',
  'libgdbm-compat-dev',
  'libssl-dev',
  'curl',
  'llvm',
  'libncurses5-dev',
  'xz-utils',
  'tk-dev',
  'liblzma-dev',
  'gdb',
  'lcov',
  'pkg-config',
  'libffi-dev',
  'libreadline6-dev',
  'lzma',
  'lzma-dev',
  'uuid-dev'
];

export {ubuntuDependencies};

export type sslUrl = {
  name: string;
  url: string;
};

const ssl102Url: sslUrl = {
  name: 'openssl@1.0',
  url: 'https://raw.githubusercontent.com/rbenv/homebrew-tap/e472b7861b49cc082d1db0f66f265368da107589/Formula/openssl%401.0.rb'
};

export {ssl102Url};
