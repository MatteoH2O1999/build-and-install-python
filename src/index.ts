// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
import * as core from '@actions/core';

core.info(`Requested Python version: ${core.getInput('python-version')}`);
core.info(`Requested architecture: ${core.getInput('architecture')}`);
core.info(`Requested cache: ${core.getInput('cache-build')}`);
core.info(`Requested build behavior: ${core.getInput('allow-build')}`);
