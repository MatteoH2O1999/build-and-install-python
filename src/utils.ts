// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

/* eslint-disable no-restricted-imports */

import {Abortable} from 'node:events';
import fs from 'fs';

export async function readdir(
  path: fs.PathLike,
  options?:
    | (fs.ObjectEncodingOptions & {withFileTypes?: false | undefined})
    | BufferEncoding
    | null
): Promise<string[]> {
  return await fs.promises.readdir(path, options);
}

export async function exists(path: fs.PathLike): Promise<boolean> {
  return fs.existsSync(path);
}

export async function symlink(
  target: fs.PathLike,
  path: fs.PathLike,
  type?: string | null | undefined
): Promise<void> {
  return await fs.promises.symlink(target, path, type);
}

export async function readFile(
  path: fs.PathLike | fs.promises.FileHandle,
  options?:
    | ({
        encoding: BufferEncoding;
        flag?: fs.OpenMode | undefined;
      } & Abortable)
    | BufferEncoding
): Promise<string> {
  const content = await fs.promises.readFile(path, options);
  if (content instanceof Buffer) {
    return content.toString();
  }
  return content;
}

export async function writeFile(
  file: fs.PathLike,
  data: string
): Promise<void> {
  return await fs.promises.writeFile(file, data);
}

export async function realpath(
  path: fs.PathLike,
  options?: BufferEncoding | fs.ObjectEncodingOptions | null
): Promise<string> {
  return await fs.promises.realpath(path, options);
}

export function realpathSync(
  path: fs.PathLike,
  options?: fs.EncodingOption
): string {
  return fs.realpathSync(path, options);
}
