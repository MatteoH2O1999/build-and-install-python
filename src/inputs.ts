// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as core from '@actions/core';
import {InputNames} from './constants';
import {getVersionInputFromFile} from 'setup-python/src/utils';
import os from 'os';
import semverValidRange from 'semver/ranges/valid';

export enum PythonType {
  CPython = 'cpython',
  PyPy = 'pypy'
}

export class PythonVersion {
  readonly type: PythonType;
  readonly version: string;

  constructor(pythonVersion: string) {
    core.debug(`Parsing version string "${pythonVersion}"...`);
    pythonVersion = pythonVersion.toLowerCase().trim();
    if (pythonVersion.length === 0) {
      pythonVersion = 'x';
    }
    while (pythonVersion.split('.').length < 3) {
      pythonVersion = pythonVersion.concat('.x');
    }
    core.debug(`Full semver range string: "${pythonVersion}".`);
    core.debug('Checking Python type (CPython or PyPy).');
    if (pythonVersion.includes('pypy')) {
      this.type = PythonType.PyPy;
      pythonVersion = pythonVersion.replace('pypy', '').replace('-', '');
    } else {
      this.type = PythonType.CPython;
    }
    core.debug(`Python type: "${this.type}".`);
    if (pythonVersion.startsWith('v')) {
      pythonVersion = pythonVersion.replace('v', '');
    }
    if (semverValidRange(pythonVersion) === null) {
      throw new Error(
        `An invalid semver string was supplied. Got "${pythonVersion}".`
      );
    }
    this.version = pythonVersion;
    core.debug(`Final resolved version range: "${this.version}".`);
  }
}

export enum BuildBehavior {
  Allow = 'allow',
  Info = 'info',
  Warn = 'warn',
  Error = 'error',
  Force = 'force'
}

export type ActionInputs = {
  version: PythonVersion;
  architecture: string;
  cache: boolean;
  buildBehavior: BuildBehavior;
  token: string;
  checkLatest: boolean;
  allowPrereleases: boolean;
};

async function getBehavior(): Promise<BuildBehavior> {
  const behaviorInput = core
    .getInput(InputNames.ALLOW_BUILD)
    .toLowerCase()
    .trim();
  core.debug(`Parsing build behavior string ${behaviorInput}...`);
  if (behaviorInput === 'error') {
    core.debug('Found behavior "error".');
    return BuildBehavior.Error;
  }
  if (behaviorInput === 'warn') {
    core.debug('Found behavior "warn".');
    return BuildBehavior.Warn;
  }
  if (behaviorInput === 'info') {
    core.debug('Found behavior "info".');
    return BuildBehavior.Info;
  }
  if (behaviorInput === 'allow') {
    core.debug('Found behavior "allow".');
    return BuildBehavior.Allow;
  }
  if (behaviorInput === 'force') {
    core.debug('Found behavior "force".');
    return BuildBehavior.Force;
  }
  throw new Error(
    `Unrecognized input value for "${
      InputNames.ALLOW_BUILD
    }". Expected one of "${BuildBehavior.Allow}", "${BuildBehavior.Info}", "${
      BuildBehavior.Warn
    }", "${BuildBehavior.Error}, ${BuildBehavior.Force}". Got "${core.getInput(
      InputNames.ALLOW_BUILD
    )}".`
  );
}

async function getCache(): Promise<boolean> {
  core.debug('Parsing cache behavior...');
  try {
    return core.getBooleanInput(InputNames.CACHE_BUILD);
  } catch (error) {
    throw new Error(
      `Expected boolean value for input "${
        InputNames.CACHE_BUILD
      }". Supported values are "true", "false", "True", "False", "TRUE", "FALSE". Got "${core.getInput(
        InputNames.CACHE_BUILD
      )}".`
    );
  }
}

async function getCheckLatest(): Promise<boolean> {
  core.debug('Parsing check latest input');
  try {
    return core.getBooleanInput(InputNames.CHECK_LATEST);
  } catch (error) {
    throw new Error(
      `Expected boolean value for input "${
        InputNames.CHECK_LATEST
      }". Supported values are "true", "false", "True", "False", "TRUE", "FALSE". Got "${core.getInput(
        InputNames.CHECK_LATEST
      )}".`
    );
  }
}

async function getAllowPrereleases(): Promise<boolean> {
  core.debug('Parsing allow prereleases input');
  try {
    return core.getBooleanInput(InputNames.PRERELEASES);
  } catch (error) {
    throw new Error(
      `Expected boolean value for input "${
        InputNames.PRERELEASES
      }". Supported values are "true", "false", "True", "False", "TRUE", "FALSE". Got "${core.getInput(
        InputNames.PRERELEASES
      )}".`
    );
  }
}

async function getArchitecture(): Promise<string> {
  const archString: string = core.getInput(InputNames.ARCHITECTURE);
  core.debug(`Parsing architecture string "${archString}"`);
  if (archString === '') {
    core.debug("architecture string is ''. Using os.arch()...");
    return os.arch();
  }
  core.debug(`Using architecture "${archString}"...`);
  return archString;
}

async function extractPythonVersion(): Promise<string> {
  let pythonVersions: string[] = core.getMultilineInput(
    InputNames.PYTHON_VERSION
  );
  let pythonVersionFile: string = core.getInput(InputNames.PYTHON_VERSION_FILE);
  core.debug(
    `Parsing python-version string "${pythonVersions.join(
      ' '
    )}" and python-version-file string "${pythonVersionFile}"`
  );

  if (pythonVersions.length > 0 && pythonVersionFile !== '') {
    core.warning(
      `Both "${InputNames.PYTHON_VERSION}" and "${InputNames.PYTHON_VERSION_FILE}" inputs were supplied. Only "${InputNames.PYTHON_VERSION}" will be used.`
    );
  }

  if (pythonVersions.length === 0 && pythonVersionFile === '') {
    core.info(
      `Neither "${InputNames.PYTHON_VERSION}" and "${InputNames.PYTHON_VERSION_FILE}" inputs were supplied. Defaulting to using ".python-version" file...`
    );
    pythonVersionFile = '.python-version';
  }

  if (pythonVersions.length === 0) {
    try {
      pythonVersions = getVersionInputFromFile(pythonVersionFile);
    } catch (error) {
      if (error instanceof Error) {
        core.warning(error.message);
      }
      pythonVersions = [];
    }
  }

  let pythonVersion = '';

  if (pythonVersions.length > 0) {
    if (pythonVersions.length > 1) {
      core.warning(
        `This action does not support installing multiple versions. Will instll only ${pythonVersions[0]}`
      );
    }
    pythonVersion = pythonVersions[0];
  }

  core.debug(`Using python version string "${pythonVersion}"...`);
  return pythonVersion;
}

export async function parseInputs(): Promise<ActionInputs> {
  return {
    allowPrereleases: await getAllowPrereleases(),
    architecture: await getArchitecture(),
    buildBehavior: await getBehavior(),
    cache: await getCache(),
    checkLatest: await getCheckLatest(),
    token: core.getInput(InputNames.TOKEN),
    version: new PythonVersion(await extractPythonVersion())
  };
}
