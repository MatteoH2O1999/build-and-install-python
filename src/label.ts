// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as core from '@actions/core';
import {getOSInfo} from 'setup-python/src/utils';
import {testedLabels} from './constants';

export async function emitWarnings(): Promise<void> {
  const osInfo = await getOSInfo();
  if (osInfo === undefined) {
    core.debug(
      'Impossible to gather information about current platform. Emitting no warnings...'
    );
    return;
  }
  const platform = osInfo.osName.toLowerCase();
  const version =
    platform === 'macos' ? osInfo.osVersion.split('.')[0] : osInfo.osVersion;
  const label = `${platform}-${version}`;
  core.debug(`Detected label: "${label}"`);
  if (!testedLabels.includes(label)) {
    core.warning(
      `This action has been tested with labels ${testedLabels.join(
        ', '
      )}. Detected label: ${label}. Action is not guaranteed to work`
    );
  }
}
