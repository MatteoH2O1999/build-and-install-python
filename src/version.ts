// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2025 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as core from '@actions/core';
import * as tc from '@actions/tool-cache';
import {ActionInputs, PythonType, PythonVersion} from './inputs';
import {ManifestUrl, minSupportedVersion} from './constants';
import path from 'path';
import semver from 'semver';

export async function isPyPy(version: PythonVersion): Promise<boolean> {
  core.debug('Checking if version is PyPy.');
  return version.type === PythonType.PyPy;
}

export async function isGraalPy(version: PythonVersion): Promise<boolean> {
  core.debug('Checking if version is GraalPy.');
  return version.type === PythonType.GraalPy;
}

export async function isCpython(version: PythonVersion): Promise<boolean> {
  core.debug('checking if version is CPython');
  return version.type === PythonType.CPython;
}

export type SetupPythonResult = {
  version: string;
  freethreaded: boolean;
  success: boolean;
};

export async function getSetupPythonResult(
  inputs: ActionInputs
): Promise<SetupPythonResult> {
  let resultVersionString: string;
  let success: boolean;
  let freethreaded: boolean;
  if (await isPyPy(inputs.version)) {
    core.debug('Version is PyPy.');
    success = true;
    freethreaded = false;
    resultVersionString = inputs.version.version;
  } else if (await isGraalPy(inputs.version)) {
    core.debug('Version is GraalPy.');
    success = true;
    freethreaded = false;
    resultVersionString = inputs.version.version;
  } else {
    core.debug('Version is CPython.');
    const arch = inputs.version.freethreaded
      ? inputs.architecture + '-freethreaded'
      : inputs.architecture;
    freethreaded = inputs.version.freethreaded;
    core.debug(`Manifest architecture: ${arch}`);
    core.debug('Checking local tool cache...');
    const localPath = tc.find('Python', inputs.version.version, arch);
    if (localPath !== '' && !inputs.checkLatest) {
      const localVersion = path.basename(path.dirname(localPath));
      resultVersionString = localVersion;
      success = true;
      core.debug(`CPython version resolved to "${resultVersionString}"`);
    } else {
      core.debug('Downloading manifest...');
      const manifest = await tc.getManifestFromRepo(
        ManifestUrl.OWNER,
        ManifestUrl.REPO,
        `token ${inputs.token}`,
        ManifestUrl.BRANCH
      );
      core.debug(
        `Checking manifest for version "${inputs.version.version}" and arch "${arch}"...`
      );
      const matchVersion = await tc.findFromManifest(
        inputs.version.version,
        false,
        manifest,
        arch
      );
      if (matchVersion === undefined) {
        const minVersion = semver.minVersion(inputs.version.version);
        if (
          inputs.allowPrereleases &&
          minVersion &&
          !semver.intersects(
            inputs.version.version,
            `>=${minVersion.major}.${minVersion.minor + 1}.0-0`
          ) &&
          semver.intersects(
            inputs.version.version,
            `>=${minVersion.major}.${minVersion.minor}.${
              minVersion.patch + 1
            }-0`
          )
        ) {
          core.debug('Testing for prerelease versions');
          const preReleaseVersion = `~${minVersion.major}.${minVersion.minor}.0-0`;
          const matchPreRelease = await tc.findFromManifest(
            preReleaseVersion,
            false,
            manifest,
            arch
          );
          if (matchPreRelease === undefined) {
            success = false;
            resultVersionString = '';
            core.debug('Could not find specified version in manifest.');
          } else {
            success = true;
            resultVersionString = matchPreRelease.version;
            core.debug(
              `CPython version resolved to prerelease ${resultVersionString}`
            );
          }
        } else {
          success = false;
          resultVersionString = '';
          core.debug('Could not find specified version in manifest.');
        }
      } else {
        resultVersionString = matchVersion.version;
        success = true;
        core.debug(`CPython version resolved to "${resultVersionString}"`);
      }
    }
  }
  return {
    freethreaded,
    success,
    version: resultVersionString
  };
}

export async function checkMinVersion(specificVersion: string): Promise<void> {
  if (semver.lt(specificVersion, minSupportedVersion)) {
    core.warning(
      `Specific version ${specificVersion} is no longer officially supported. This action will not break support intentionally but no effort will be spent in supporting it.`
    );
  }
}
