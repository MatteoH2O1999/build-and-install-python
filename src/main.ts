// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as core from '@actions/core';
import * as tc from '@actions/tool-cache';
import {ActionInputs, BuildBehavior, parseInputs} from './inputs';
import {InputNames, OutputNames} from './constants';
import {SetupPythonResult, getSetupPythonResult, isPyPy} from './version';
import getBuilder from './builder';

export default async function main(): Promise<void> {
  let inputs: ActionInputs;
  let setupPythonResult: SetupPythonResult;

  // Input parsing

  core.debug('Parsing inputs...');
  try {
    inputs = await parseInputs();
  } catch (error) {
    core.setOutput(OutputNames.PYTHON_VERSION, '');
    core.setOutput(OutputNames.ARCHITECTURE, '');
    let message = 'Error while parsing inputs.';
    if (error instanceof Error) {
      message = message.concat('\n').concat(error.message);
    }
    core.setFailed(message);
    return;
  }
  core.debug('Inputs successfully parsed.');

  core.debug(
    `Requested python version: ${inputs.version.type} ${inputs.version.version}`
  );
  core.debug(`Requested architecture: ${inputs.architecture}`);
  core.debug(`Requested build cache: ${inputs.cache}`);
  core.debug(
    `Requested behavior for deprecated builds: ${inputs.buildBehavior}`
  );

  core.setOutput(OutputNames.ARCHITECTURE, inputs.architecture);

  // Checking if actions/setup-python fails

  core.debug('Resolving setup-python version...');
  try {
    setupPythonResult = await getSetupPythonResult(inputs);
  } catch (error) {
    let version = '';
    if (isPyPy(inputs.version)) {
      version = 'pypy';
    }
    core.setOutput(
      OutputNames.PYTHON_VERSION,
      version.concat(inputs.version.version)
    );
    let message = 'Error while resolving setup-python version.';
    if (error instanceof Error) {
      message = message.concat('\n').concat(error.message);
    }
    core.setFailed(message);
    return;
  }
  core.debug('setup-python version resolved.');

  if (setupPythonResult.success) {
    // actions/setup-python already supports the version: doing nothing

    if (isPyPy(inputs.version)) {
      core.info(
        `PyPy version ${inputs.version.version} is supported by actions/setup-python with specific version ${setupPythonResult.version}.`
      );
    } else {
      core.info(
        `CPython version ${inputs.version.version} is supported by actions/setup-python with specific version ${setupPythonResult.version}.`
      );
    }
    core.setOutput(OutputNames.PYTHON_VERSION, setupPythonResult.version);
    return;
  } else {
    if (isPyPy(inputs.version)) {
      // This action does not support building PyPy

      core.info(
        `PyPy version ${inputs.version.version} is not supported by actions/setup-python.`
      );
      core.setFailed('This action does not support building PyPy from source');
      core.setOutput(OutputNames.PYTHON_VERSION, '');
      return;
    }
    core.info(
      `CPython version ${inputs.version.version} is not supported by actions/setup-python.`
    );

    // Apply inputs.buildBehavior

    switch (inputs.buildBehavior) {
      case BuildBehavior.Error:
        core.info(
          'Requested behavior for deprecated builds: error. Failing...'
        );
        core.setFailed(
          `CPython version ${inputs.version.version} is not supported by actions/setup-python.
          If you wish for this action to build CPython ${inputs.version.version} from source set input "${InputNames.ALLOW_BUILD}" to
          either "${BuildBehavior.Warn}", "${BuildBehavior.Info}" or "${BuildBehavior.Allow}".`
        );
        return;
      case BuildBehavior.Warn:
        core.warning(
          `CPython version ${inputs.version.version} is not supported by actions/setup-python.
          This probably means you are using a deprecated version. If this is not the case, you may suppress the warning setting input "${InputNames.ALLOW_BUILD}"
          to either "${BuildBehavior.Info}" or "${BuildBehavior.Allow}".`
        );
        break;
      case BuildBehavior.Info:
        core.info(
          `CPython version ${inputs.version.version} will be built from source.`
        );
        break;
      case BuildBehavior.Allow:
        core.debug(
          `CPython version ${inputs.version.version} will be built from source.`
        );
        break;
    }

    // Create builder

    const builder = await getBuilder(
      inputs.version,
      inputs.architecture,
      inputs.token
    );

    // If builder is null, the version cannot be built from source

    if (builder === null) {
      core.setFailed(
        `CPython version ${inputs.version.version} cannot be built from source.`
      );
      return;
    }

    let buildPath: string | null = null;

    // Try to restore cache

    if (inputs.cache) {
      try {
        buildPath = await builder.restoreCache();
      } catch (error) {
        const message = 'Error while restoring cache.';
        if (error instanceof Error) {
          message.concat('\n').concat(error.message);
        }
        core.setFailed(message);
        return;
      }
    }

    if (buildPath != null) {
      // Cache-hit
      core.info('Cache-hit. Copying already built version to tool cache');
      await tc.cacheDir(
        buildPath,
        'Python',
        builder.specificVersion,
        builder.arch
      );
      core.info(
        `CPython ${builder.specificVersion} for arch ${builder.arch} successfully installed.`
      );
      core.setOutput(OutputNames.PYTHON_VERSION, builder.specificVersion);
      await builder.clean();
      return;
    }

    // Cache miss or not used

    try {
      buildPath = await builder.build();
    } catch (error) {
      const message = 'Error while building Python.';
      if (error instanceof Error) {
        message.concat('\n').concat(error.message);
      }
      core.setFailed(message);
      return;
    }

    // Save cache

    if (inputs.cache) {
      try {
        await builder.saveCache();
      } catch (error) {
        const message = 'Error while saving cache.';
        if (error instanceof Error) {
          message.concat('\n').concat(error.message);
        }
        core.setFailed(message);
        return;
      }
    }

    // Install in tool cache and clean build folder

    core.info('Copying built folder into tool cache.');
    await tc.cacheDir(
      buildPath,
      'Python',
      builder.specificVersion,
      builder.arch
    );
    core.info(
      `CPython ${builder.specificVersion} for arch ${builder.arch} successfully installed.`
    );
    await builder.clean();
    core.setOutput(OutputNames.PYTHON_VERSION, builder.specificVersion);
    return;
  }
}
