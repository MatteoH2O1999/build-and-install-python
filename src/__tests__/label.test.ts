// Action to build any Python version on the latest labels and install it into the local tool cache.
// Copyright (C) 2022 Matteo Dell'Acqua
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

import * as constants from '../constants';
import * as core from '@actions/core';
import * as label from '../label';
import * as utils from 'setup-python/src/utils';
import {describe, expect, jest, test} from '@jest/globals';

jest.mock('setup-python/src/utils');
jest.mock('@actions/core');

const mockedUtils = jest.mocked(utils);
const mockedCore = jest.mocked(core);

//@ts-expect-error overwrite of const property for testing
constants.testedLabels = ['macos-12', 'ubuntu-22.04', 'windows-2022'];

describe('emitWarnings', () => {
  test('calls getOSInfo', async () => {
    await label.emitWarnings();

    expect(mockedUtils.getOSInfo).toBeCalledTimes(1);
  });

  test('logs impossibility to gather information in core.debug', async () => {
    mockedUtils.getOSInfo.mockResolvedValueOnce(undefined);

    await label.emitWarnings();

    expect(mockedCore.debug.mock.calls).toMatchSnapshot();
  });

  test.each([
    {osName: 'Ubuntu', osVersion: '20.04'},
    {osName: 'Macos', osVersion: '13.4.2'}
  ])(
    'emit warning if label is not tested with osName: "$osName" and osVersion: "$osVersion"',
    async detectedOs => {
      mockedUtils.getOSInfo.mockResolvedValueOnce(detectedOs);

      await label.emitWarnings();

      expect(mockedCore.warning.mock.calls).toMatchSnapshot();
    }
  );

  test.each([
    {osName: 'Ubuntu', osVersion: '22.04'},
    {osName: 'Macos', osVersion: '12.4.2'}
  ])(
    'do not emit warning if label is not tested with osName: "$osName" and osVersion: "$osVersion"',
    async detectedOs => {
      mockedUtils.getOSInfo.mockResolvedValueOnce(detectedOs);

      await label.emitWarnings();

      expect(mockedCore.warning).not.toBeCalled();
    }
  );
});
