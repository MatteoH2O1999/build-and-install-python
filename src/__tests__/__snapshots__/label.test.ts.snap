// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emitWarnings emit warning if label is not tested with osName: "Macos" and osVersion: "13.4.2" 1`] = `
[
  [
    "This action has been tested with labels macos-12, ubuntu-22.04, windows-2022. Detected label: macos-13. Action is not guaranteed to work",
  ],
]
`;

exports[`emitWarnings emit warning if label is not tested with osName: "Ubuntu" and osVersion: "20.04" 1`] = `
[
  [
    "This action has been tested with labels macos-12, ubuntu-22.04, windows-2022. Detected label: ubuntu-20.04. Action is not guaranteed to work",
  ],
]
`;

exports[`emitWarnings logs impossibility to gather information in core.debug 1`] = `
[
  [
    "Impossible to gather information about current platform. Emitting no warnings...",
  ],
]
`;
